### PR TITLE
feat(ELY-397): Refactor external-roles module

### PR DESCRIPTION
**Pull request by askoriy**
This PR provides refactored `external-project-iam-roles` module, which now is flexible enough to be also used directly.
For shorter external reference and because breaking changes the module was renamed to `external-roles`.

Note: During applying this module all roles will be recreated with potentially fail on the first run because delete/create simultaneously. The second run `terraform apply` must fix breaking state

### CODE REVIEW

File: examples/external-roles/main.tf [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    "terraform@tf-root-project.iam.gserviceaccount.com" = {
+      "ecosystem-staging-ec90" = [
+        "role/editor",
+      ],
+      "ecosystem-cnp-staging-e1f7" = [
+        "roles/compute.admin",
+      ]
+    }
+  }
+}
```
Comment (anakovt) [inline examples/external-roles/main.tf:14]: https://github.com/extenda/tf-module-gcp-project/blob/11250639a16c0d7192229adc0669fde949a68dd0/examples/external-roles/provider.tf#L11-L21
why do you need this block for the example?

File: main.tf [hunk 2]
```diff
@@ -85,23 +85,50 @@ module "services_sa" {
...
+    for service in var.services:
+      "${service.name}@${module.project_factory.project_id}.iam.gserviceaccount.com" => {
+        "${var.parent_project_id}" = var.common_iam_roles
+      }
 
-  project_id       = module.project_factory.project_id
-  services         = var.services
-  common_iam_roles = var.common_iam_roles
+  }
   sa_depends_on    = module.services_sa.email
```
Comment (anakovt) [inline main.tf]: terraform fmt

File: main.tf [hunk 3]
```diff
@@ -85,23 +85,50 @@ module "services_sa" {
...
+
+module "parent_project_roles" {
+  source = "./modules/external-roles"
+
+  roles_map = {
+    "${local.ci_cd_sa_email}" = {
+      "${var.parent_project_id}" = var.parent_project_iam_roles
+    }
+  }
+  sa_depends_on    = module.services_sa.email
```
Comment (anakovt) [inline main.tf]: terraform fmt

File: main.tf [hunk 4]
```diff
@@ -85,23 +85,50 @@ module "services_sa" {
...
-  dns_project_id        = var.dns_project_id
-  dns_project_iam_roles = var.dns_project_iam_roles
-  gcr_project_id        = var.gcr_project_id
-  gcr_project_iam_roles = var.gcr_project_iam_roles
+  roles_map = {
+    "${local.ci_cd_sa_email}" = {
+      "${var.dns_project_id}" = var.dns_project_iam_roles
+    }
+  }
+  sa_depends_on    = module.services_sa.email
```
Comment (anakovt) [inline main.tf]: terraform fmt

File: main.tf [hunk 5]
```diff
@@ -85,23 +85,50 @@ module "services_sa" {
...
+
+module "gcr_project_roles" {
+  source = "./modules/external-roles"
+
+  roles_map = {
+    "${local.ci_cd_sa_email}" = {
+      "${var.gcr_project_id}" = var.gcr_project_iam_roles
+    }
+  }
+  sa_depends_on    = module.services_sa.email
```
Comment (anakovt) [inline main.tf]: terraform fmt

File: modules/external-roles/README.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
+This module is for creating specific iam roles in specific projects for existing service account.
```
Comment (anakovt) [inline modules/external-roles/README.md]: `This module is intended to create specific IAM roles in specific projects for the existing service account.`

File: modules/external-roles/README.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+Module accept a list of service accounts, projects and roles in next data structure passed to roles_map variable:
```
Comment (anakovt) [inline modules/external-roles/README.md]: The module accepts..... and roles. The following data structure is passed to roles_map variable:

File: modules/external-roles/vars.tf [hunk 8]
```diff
@@ -0,0 +1,10 @@
+variable roles_map {
+  description = "Map of roles in external projects"
+  type = map(map(list(string)))
+}
+
+variable sa_depends_on {
+  description = "Service Account which this module depends on"
+  type        = any
+  default     = ""
+}
```
Comment (anakovt) [inline modules/external-roles/vars.tf]: new line

**Review by anakovt — COMMENTED** [review]
see comments

**Comment by akovtun-extenda** [comment]
Another question: Let's imagine that the project `experience-selfscan-staging` has `mcpe-auth` and `mcpe-selfscan` SAs that need different roles in different projects.  mcpe-auth - roles/storage.admin in the `extenda` project, and `mcpe-selfscan` - roles/monitoring.editor in the `experience-staging`. How will we accomplish this? As far as I understood you can have only one `roles_map` per one calling module.

**Pull request merged**